python/ec2-exposed-group.py

```diff
@@ -0,0 +1,103 @@
+#
+# This file made available under CC0 1.0 Universal (https://creativecommons.org/publicdomain/zero/1.0/legalcode)
+#
+# Ensure that no security groups allow public access to the specified ports.
+# Description: Checks that all security groups block access to the specified ports.
+#
+# Trigger Type: Change Triggered
+# Scope of Changes: EC2:SecurityGroup
+# Accepted Parameters: examplePort1, exampleRange1, examplePort2, ...
+# Example Value: 8080, 1-1024, 2375, ...
+
+
+import json
+import boto3
+
+
+APPLICABLE_RESOURCES = ["AWS::EC2::SecurityGroup"]
+
+
+def expand_range(ports):
+    if "-" in ports:
+        return range(int(ports.split("-")[0]), int(ports.split("-")[1]))
+    else:
+        return int(ports)
+
+
+def find_exposed_ports(ip_permissions):
+    exposed_ports = []
+    for permission in ip_permissions:
+        if next((r for r in permission["IpRanges"]
+                if "0.0.0.0/0" in r["CidrIp"]), None):
+                    exposed_ports.extend(range(permission["FromPort"],
+                                               permission["ToPort"]))
+    return exposed_ports
+
+
+def find_violation(ip_permissions, forbidden_ports):
+    exposed_ports = find_exposed_ports(ip_permissions)
+    for forbidden in forbidden_ports:
+        ports = expand_range(forbidden_ports[forbidden])
+        for port in ports:
+            if port in exposed_ports:
+                return "A forbidden port is exposed to the internet."
+            else:
+                return None
+
+
+def evaluate_compliance(configuration_item, rule_parameters):
+    if configuration_item["resourceType"] not in APPLICABLE_RESOURCES:
+        return {
+            "compliance_type": "NOT_APPLICABLE",
+            "annotation": "The rule doesn't apply to resources of type " +
+            configuration_item["resourceType"] + "."
+        }
+
+    ec2 = boto3.resource("ec2")
+    security_group = ec2.SecurityGroup(configuration_item["resourceId"])
```

Comment:
Thank you very much for your suggestion, it will be very useful in the future. I have updated both exposed rules to use the configuration attribute and also the MFA rule, which now doesn't have to rely on a hack to obtain the username of the resource.